src/claim_canister/src/lib.rs

```diff
@@ -1,41 +1,141 @@
+    let ledger_id = LEDGER_CANISTER_ID.with(|cell| {
+        let id = cell.borrow().get().clone();
+        if id == Principal::anonymous() {
+            return Err(ClaimError::LedgerNotSet);
+        }
+        Ok(id)
+    })?;
+
+    let account = Account {
+        owner: ic_cdk::api::id(),
+        subaccount: Some(subaccount),
+    };
+
+    let result: Result<(Nat,), _> = call(ledger_id, "icrc1_balance_of", (account,)).await;
+
+    match result {
+        Ok((balance,)) => Ok(balance),
+        Err(_) => Err(ClaimError::DepositTransferFailed),
+    }
+}
+
+async fn transfer_icrc1(
+    from_subaccount: Option<Vec<u8>>,
+    to: Principal,
+    gross_amount: Nat,
+) -> Result<(), ClaimError> {
+    let ledger_id = LEDGER_CANISTER_ID.with(|cell| {
+        let id = cell.borrow().get().clone();
+        if id == Principal::anonymous() {
+            return Err(ClaimError::LedgerNotSet);
+        }
+        Ok(id)
+    })?;
+
+    if gross_amount <= TRANSFER_FEE.clone() {
+        return Ok(());
+    }
+
+    let net_amount = gross_amount - TRANSFER_FEE.clone();
+
+    let transfer_args = TransferArg {
+        from_subaccount,
+        to: Account {
+            owner: to,
+            subaccount: None,
+        },
+        amount: net_amount,
+        fee: Some(TRANSFER_FEE.clone()),
+        memo: None,
+        created_at_time: None,
+    };
+
+    let result: Result<(Result<Nat, TransferError>,), _> =
+        call(ledger_id, "icrc1_transfer", (transfer_args,)).await;
+
+    match result {
+        Ok((Ok(_),)) => Ok(()),
+        _ => Err(ClaimError::DepositTransferFailed),
+    }
+}
 
 #[ic_cdk::init]
-pub fn init(owner: Principal) {
+pub fn init(owner: Principal, claim_deposit: Nat, ledger_canister_id: Principal, approval_period: u64) {
     OWNER.with(|cell| {
         cell.borrow_mut().set(owner).ok();
     });
 
     APPROVERS.with(|approvers| {
         approvers.borrow_mut().insert(owner, true);
     });
+
+    CLAIM_DEPOSIT.with(|cell| {
+        cell.borrow_mut().set(StorableNat(claim_deposit)).ok();
+    });
+
+    LEDGER_CANISTER_ID.with(|cell| {
+        cell.borrow_mut().set(ledger_canister_id).ok();
+    });
+
+    APPROVAL_PERIOD.with(|cell| {
+        cell.borrow_mut().set(approval_period).ok();
+    });
 }
 
 #[ic_cdk::update]
-pub fn add_claim(
+pub async fn add_claim(
     receiver: Principal,
     amount: Nat,
     pool_canister_id: Principal,
     description: String,
 ) -> Result<u64, ClaimError> {
+    let caller = ic_cdk::api::caller();
+
     let claim_id = CLAIM_COUNTER.with(|counter| {
         let current = counter.borrow().get().clone();
         let new_counter = current + 1;
         counter.borrow_mut().set(new_counter).ok();
         new_counter
     });
 
+    let required_deposit = CLAIM_DEPOSIT.with(|cell| cell.borrow().get().clone().0);
+
+    if required_deposit > Nat::from(0u64) {
+        let subaccount = get_claim_deposit_subaccount(caller, claim_id);
```

Comment:
Check out my initial comment.
I said the claim number created by a user.
What if we have two users who are trying to create a deposit at the same time?